### PR TITLE
feat: throw SERVICE_CURRENTLY_UNAVAILABLE in Producer if the lag is above the configured maxLag

### PR DESCRIPTION
**Pull request by kalinkrustev**
This introduces a new property `lagMonitor` in the `Producer` configuration. It is an object having the following properties:
- `topic` - topic for which to monitor the lag
- `consumerGroup` - consumer group for which to monitor the lag
- `interval` - an interval in seconds for querying Kafka for the current lag
- `max` - a limit over which producing messages with `sendMessage` will throw an error

This is useful to avoid overloading the system with more messages than it is sized to handle, leading to too many cases of time outs.

### CODE REVIEW

**Comment by bushjames** [comment]
I love this feature and I think it deserves a discussion in the DA (about back-pressure in general) @elnyry-sam-k @MichaelJBRichards .

**Comment by bushjames** [comment]
@kalinkrustev It is not clear to me how you choose which consumers to monitor lag on? Each consumer will be at a different offset, some might be keeping up and some might be falling behind...you might care about some but not all of them. Please could you explain a bit how you see this working in practice? Many thanks.

**Comment by kalinkrustev** [comment]
@bushjames thanks for your question, you are absolutely right to ask that. It is very strange that this method `queryWatermarkOffsets` works in the Producer and returns data. Maybe it is a bit misleading, but let me check about that. The lags should be per some consumer group, so probably I need to pass this somewhere.

**Comment by kalinkrustev** [comment]
It seems `queryWatermarkOffsets` returns the offsets of the messages retained in the partitions. It is not very well explained in the documentation. I will try to find if we have some method to obtain the current offsets for a consumer group.

**Comment by kalinkrustev** [comment]
@bushjames I have found the method I needed: [consumer.committed()](https://github.com/Blizzard/node-rdkafka/blob/a950a56a43d679176209b43ed3e5817d0c1097ac/lib/kafka-consumer.js#L200) . It was not documented in the readme. I have also updated PR the description to mention the consumer group.

So when the lag monitoring is active, we create a consumer, which is manually assigned to all partitions, without consuming anything. This manual assignment excludes it from the group auto-assignment logic, as mentioned in the [Kafka docs](https://kafka.apache.org/22/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#assign-java.util.Collection-)). This allows us to call the committed() method and obtain the current offsets for the consumer group.

**Review by vijayg10 — APPROVED** [review]

File: src/kafka/producer.js [hunk 1]
```diff
@@ -479,6 +499,65 @@ class Producer extends EventEmitter {
...
+    }, {
+      'auto.offset.reset': 'earliest'
+    })
+    await new Promise((resolve, reject) => {
+      this._consumer.connect({ topic: this._config.lagMonitor.topic }, (err, metadata) => {
+        if (err) {
+          reject(err)
+        } else {
+          const partitionIds = metadata.topics.find(topic => topic.name === this._config.lagMonitor.topic).partitions.map(partition => partition.id)
+          this._consumer.assign(partitionIds.map(partition => ({ topic: this._config.lagMonitor.topic, partition })))
```
Comment (geka-evk) [inline src/kafka/producer.js:554]: Won't it cause re-balancing for consumers from `consumerGroup`?

**Pull request merged**